Sources/Auth/Internal/AutoRefreshToken.swift

```diff
@@ -0,0 +1,76 @@
+//
+//  AutoRefreshToken.swift
+//
+//
+//  Created by Guilherme Souza on 06/04/24.
+//
+
+import _Helpers
+import ConcurrencyExtras
+import Foundation
+
+actor AutoRefreshToken {
+  /// Current session will be checked for refresh at this interval.
+  static let tickDuration: TimeInterval = 30
+  /// A token refresh will be attempted this many ticks before the current session expires.
+  static let tickThreshold = 3
+
+  private var task: Task<Void, Never>?
+
+  @Dependency(\.sessionManager) var sessionManager
+  @Dependency(\.logger) var logger
+
+  func start() {
+    logger?.debug("")
+
+    task?.cancel()
+    task = Task {
+      while !Task.isCancelled {
+        await autoRefreshTokenTick()
+        try? await Task.sleep(nanoseconds: UInt64(AutoRefreshToken.tickDuration) * NSEC_PER_SEC)
```

Comment:
I was curious what your thoughts were on using this Task.sleep to precisely schedule the time that the token should be refreshed instead of checking every 30 seconds. This would match the behavior of the kotlin library.

Here is an example of an implementation: https://github.com/supabase/supabase-swift/pull/391/commits/f62567e90ef2754721728c26042b5d08024be1bc

```
 func scheduleSessionRefresh(_ session: Session) throws {
    if autoRefreshTask != nil {
      return
    }

    autoRefreshTask = Task {
      defer { autoRefreshTask = nil }

      guard let expiresAt = session.expiresAt else {
        return
      }
      let expiryDate = Date(timeIntervalSince1970: expiresAt)

      let timeIntervalToExpiry = expiryDate.timeIntervalSinceNow

      // if negative then token is expired and will refresh right away
      let timeIntervalToRefresh = max(timeIntervalToExpiry * 0.8, 0)

      try await Task.sleep(nanoseconds: UInt64(timeIntervalToRefresh * 1_000_000_000))
      let session = try await sessionRefresher.refreshSession(session.refreshToken)
    }
  }
  ```

Reply:
Hi @JOsacky the 30-second interval follows the implementation of the official JS and Flutter libraries, but in Swift, we could do it the way you suggested.

I'd like input from @hf @J0 or @kangmingtay on this.

By checking every 30 seconds, we can detect sessions logged out early, this is the only thing I can think of that would benefit from the current implementation.

Reply:
Just seeing this comment. Thank you for thinking of the tradeoffs between the two! It seems like either would be a good solution and I saw that you chose the scheduled approach. Thank you for all the thought you put into this, I'm really excited for the next version of the swift app, I think this will greatly reduce the number of expired sessions.